starknet_py/contract.py

```diff
@@ -398,44 +386,42 @@ async def from_address(
+        :raises ProxyResolutionError: when given ProxyChecks were not sufficient to resolve proxy's implementation
         :param address: Contract's address
         :param client: Client
         :param proxy_config: Proxy resolving config
-            If set to ``True``, will use default proxy checks and :class:
-            `starknet_py.proxy_check.OpenZeppelinProxyCheck`
-            and :class:`starknet_py.proxy_check.ArgentProxyCheck` and default max_steps = 5.
+            If set to ``True``, will use default proxy checks
+            :class:`starknet_py.proxy_check.OpenZeppelinProxyCheck`
+            and :class:`starknet_py.proxy_check.ArgentProxyCheck`.
 
             If set to ``False``, :meth:`Contract.from_address` will not resolve proxies.
 
-            If a valid `ProxyConfig` is provided, will use values from that instead supplementing
-            with defaults when needed.
+            If a valid :class:`starknet_py.contract_abi_resolver.ProxyConfig` is provided, will use
+            its values instead, supplementing with defaults when needed.
 
         :return: an initialized Contract instance
         """
-        default_config: Contract.ProxyConfig = {
-            "max_steps": 5,
-            "proxy_checks": [ArgentProxyCheck(), OpenZeppelinProxyCheck()],
-        }
-        if isinstance(proxy_config, bool):
-            proxy_config = default_config if proxy_config else {}
-        else:
-            proxy_config = {**default_config, **proxy_config}
-
-        proxy_client = client.client if isinstance(client, AccountClient) else client
-        if not isinstance(proxy_client, GatewayClient):
+        address = parse_address(address)
+        proxy_config = prepare_proxy_config(proxy_config)
+        actual_client = client.client if isinstance(client, AccountClient) else client
+        if not isinstance(actual_client, GatewayClient):
+            # TODO: Add support for FullNodeClient once abi is available in RPC
             raise TypeError(
                 "Contract.from_address only supports GatewayClient or AccountClients using GatewayClient"
             )
 
-        contract = await ContractFromAddressFactory(
-            address=address,
-            client=proxy_client,
-            max_steps=proxy_config.get("max_steps", 1),
-            proxy_checks=proxy_config.get("proxy_checks", []),
-        ).make_contract()
+        direct_abi = await ContractAbiResolver.resolve_direct(
+            address=address, client=client
+        )
+        direct_contract = Contract(address=address, abi=direct_abi, client=client)
+        if not proxy_config:
+            return direct_contract
 
-        return Contract(address=address, abi=contract.data.abi, client=client)
+        abi = await ContractAbiResolver.resolve_proxy_step(
+            proxy_contract=direct_contract, proxy_config=proxy_config
+        )
+        return Contract(address=address, abi=abi, client=client)
```

Comment:
I think this flow can still be improved. Contract shouldn't need to be aware of details about proxies. Instead it should be possible to do something along the lines of `ContractAbiResolve.resolve(...)` (or even name this module something different).